File: c/secure-password-generator/generator.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void generar_contraseña(int longitud, int incluir_mayusculas, int incluir_minusculas, int incluir_digitos, int incluir_simbolos)
{
  const char mayusculas[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
  const char minusculas[] = "abcdefghijklmnopqrstuvwxyz";
  const char digitos[] = "123456789";
  const char simbolos[] = "!@#$%^&*()-_=+[]{}|;:,.<>?/";
  
  char caracteres_posibles[256] = {0};
  int idx = 0;
  
  if (incluir_mayusculas) {
    for (int i = 0; mayusculas[i] != '\0'; i++) {
      caracteres_posibles[idx++] = mayusculas[i];
    }
  }
  
  if (incluir_minusculas) {
    for (int i = 0; minusculas[i] != '\0'; i++) {
      caracteres_posibles[idx++] = minusculas[i];
    }
  }

  if (incluir_digitos) {
    for (int i = 0; digitos[i] != '\0'; i++) {
      caracteres_posibles[idx++] = digitos[i];
    }
  }

  if (incluir_simbolos) {
    for (int i = 0; simbolos[i] != '\0'; i++) {
      caracteres_posibles[idx++] = simbolos[i];
    }
  }

  if (idx == 0) {
    printf("No se seleccionaron caracteres válidos.\n");
    return;
  }
  
  srand(time(NULL));
  printf("Contraseña aleatoria: ");
    for (int i = 0; i < longitud; i++) {
        char caracter_aleatorio = caracteres_posibles[rand() % idx];
        printf("%c", caracter_aleatorio);
    }
    printf("\n");
}

int main()
{
  int longitud;
  int incluir_mayusculas, incluir_minusculas, incluir_digitos, incluir_simbolos;

  printf("¿Cuántos caracteres debe tener la contraseña? ");
  scanf("%d", &longitud);

  printf("¿Incluir mayúsculas? (1 para si, 0 para no): ");
  scanf("%d", &incluir_mayusculas);

  printf("¿Incluir minúsculas? (1 para si, 0 para no): ");
  scanf("%d", &incluir_minusculas);

  printf("¿Incluir digitos? (1 para si, 0 para no): ");
  scanf("%d", &incluir_digitos);

  printf("¿Incluir simbolos? (1 para si, 0 para no): ");
  scanf("%d", &incluir_simbolos);

  generar_contraseña(longitud, incluir_mayusculas, incluir_minusculas, incluir_digitos, incluir_simbolos);

  return 0;
}

File: python/secure-password-generator/generator.py
"""
Módulo para generar contraseñas seguras.

Funciones:
- generar_contraseña: Crea una contraseña basada en parámetros específicos (mayúsculas, minúsculas, números, caracteres especiales).
- contraseña_vocabulario: Genera contraseñas utilizando caracteres de un archivo de vocabulario.
- main: Interfaz principal para seleccionar opciones de generación de contraseñas.

Características:
- Permite la personalización de contraseñas según los tipos de caracteres.
- Incluye una opción para generar contraseñas a partir de vocabularios personalizados.
- Utiliza `tkinter` para seleccionar archivos desde una interfaz gráfica.

Dependencias:
- random y string para generar contraseñas.
- tkinter para selección de archivos.

Advertencias:
- Asegúrate de seleccionar correctamente el archivo del vocabulario si usas esa opción.
"""


import random
import string
import tkinter as tk
from tkinter import filedialog

def generar_contraseña(longitud, mayusculas, minusculas, numeros, caracteres_especiales):
    caracteres = ""
    
    if mayusculas.lower() == 's':
        caracteres += string.ascii_uppercase

    if minusculas.lower() == 's':
        caracteres += string.ascii_lowercase

    if numeros.lower() == 's':
        caracteres += string.digits

    if caracteres_especiales.lower() == 's':
        caracteres += string.punctuation

    if caracteres:
        contraseña = ''.join(random.choice(caracteres) for _ in range(longitud))
        return contraseña
    else:
        print("Debe elegir al menos un tipo de carácteres.")
        
def contraseña_vocabulario(caracteres_vocabulario, longitud):
    contraseña = ''.join(random.choice(caracteres_vocabulario) for _ in range(longitud))
    print(f"Su contraseña generada por vocabulario es: {contraseña}")

def main():   
    print("GENERADOR DE CONTRASEÑAS SEGURAS")
    print("1. Generar contraseña con carácteres")
    print("2. Generar contraseña con vocabulario")
    seleccion = input("Seleccione una opción: ")

    if seleccion == '1':

        longitud = int(input("Longitud de su contraseña: "))
        mayusculas = str(input("Mayúsculas?(s/n): "))
        minusculas = str(input("Minúsculas?(s/n): "))
        numeros = str(input("Números?(s/n): "))
        caracteres_especiales = input("Carácteres especiales?(s/n): ")
        contraseña_generada = generar_contraseña(longitud, mayusculas, minusculas, numeros, caracteres_especiales)
        print(f"Su contraseña generada es: {contraseña_generada}")
    
    if seleccion == '2':
        root = tk.Tk()
        root.withdraw()
        ruta_archivo = filedialog.askopenfilename(title="Seleccione el archivo del vocabulario")

        if ruta_archivo:
            longitud = int(input("Ingrese la longitud de su contraseña: "))
            try:
                with open(ruta_archivo, "r", encoding="utf-8") as vocabulario:
                    contenido = vocabulario.read().replace("\n", "")
                    caracteres_vocabulario = set(contenido)
                    cadena_caracteres = ''.join(caracteres_vocabulario)
                    contraseña_vocabulario(cadena_caracteres, longitud)

            except FileNotFoundError:
                print("Archivo no encontrado.")
            except FileExistsError:
                print("El archivo no existe.")
        

if __name__ == "__main__":
    main()